### ISSUE TITLE
exception while generating invoices

### ISSUE THREAD

**Issue by tytkal** [killbill/killbill#706]
I get this exception while generating invoices 

```
2017-02-20T12:16:56,558+0000 lvl='ERROR', log='CallableCallbackBase', th='bus_events-th', xff='', rId='', aRId='1', tRId='1', Fatal NotificationQ dispatch error, data corruption...
com.google.common.eventbus.EventBusException: java.lang.reflect.InvocationTargetException
	at com.google.common.eventbus.EventBusThatThrowsException.dispatchWithException(EventBusThatThrowsException.java:108)
	at com.google.common.eventbus.EventBusThatThrowsException.dispatchQueuedEventsWithException(EventBusThatThrowsException.java:94)
	at com.google.common.eventbus.EventBusThatThrowsException.postWithException(EventBusThatThrowsException.java:82)
	at org.killbill.bus.DefaultPersistentBus.dispatchBusEventWithMetrics(DefaultPersistentBus.java:299)
	at org.killbill.bus.dispatching.BusCallableCallback.dispatch(BusCallableCallback.java:40)
	at org.killbill.bus.dispatching.BusCallableCallback.dispatch(BusCallableCallback.java:29)
	at org.killbill.queue.dispatching.Dispatcher$CallableQueue.call(Dispatcher.java:103)
	at org.killbill.queue.dispatching.Dispatcher$CallableQueue.call(Dispatcher.java:86)
	at org.killbill.commons.concurrent.WrappedCallable.call(WrappedCallable.java:42)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at org.killbill.commons.concurrent.WrappedRunnable.run(WrappedRunnable.java:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.reflect.InvocationTargetException: null
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.google.common.eventbus.EventSubscriber.handleEvent(EventSubscriber.java:74)
	at com.google.common.eventbus.EventBusThatThrowsException.dispatchWithException(EventBusThatThrowsException.java:106)
	... 13 common frames omitted
Caused by: java.lang.IndexOutOfBoundsException: Index: -1, Size: 0
	at java.util.LinkedList.checkElementIndex(LinkedList.java:553)
	at java.util.LinkedList.get(LinkedList.java:474)
	at org.killbill.billing.invoice.usage.ContiguousIntervalConsumableInArrear.build(ContiguousIntervalConsumableInArrear.java:138)
	at org.killbill.billing.invoice.usage.SubscriptionConsumableInArrear.computeInArrearUsageInterval(SubscriptionConsumableInArrear.java:165)
	at org.killbill.billing.invoice.usage.SubscriptionConsumableInArrear.computeMissingUsageInvoiceItems(SubscriptionConsumableInArrear.java:114)
	at org.killbill.billing.invoice.generator.UsageInvoiceItemGenerator.generateItems(UsageInvoiceItemGenerator.java:124)
	at org.killbill.billing.invoice.generator.DefaultInvoiceGenerator.generateInvoice(DefaultInvoiceGenerator.java:85)
	at org.killbill.billing.invoice.InvoiceDispatcher.processAccountWithLockAndInputTargetDate(InvoiceDispatcher.java:313)
	at org.killbill.billing.invoice.InvoiceDispatcher.processAccountWithLock(InvoiceDispatcher.java:242)
	at org.killbill.billing.invoice.InvoiceDispatcher.processAccount(InvoiceDispatcher.java:213)
	at org.killbill.billing.invoice.InvoiceDispatcher.processSubscriptionInternal(InvoiceDispatcher.java:199)
	at org.killbill.billing.invoice.InvoiceDispatcher.processSubscriptionForInvoiceGeneration(InvoiceDispatcher.java:174)
	at org.killbill.billing.invoice.InvoiceDispatcher.processSubscriptionForInvoiceGeneration(InvoiceDispatcher.java:170)
	at org.killbill.billing.invoice.InvoiceListener.handleSubscriptionTransition(InvoiceListener.java:75)
	... 19 common frames omitted
```

this exception is blocking all future invoices for the account.
killbill version = 0.16.9

**Comment by sbrossie** [comment]
@tytkal It's indeed pretty clear from the [code](https://github.com/killbill/killbill/blob/killbill-0.16.9/invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalConsumableInArrear.java#L138) that if the `transitionTimes` list is empty we hit NPE.

It would be easy to add a check and avoid the NPE, but id like to understand better the use case (otherwise, it could break down the line). Could you provide some more info on what is happening:

Ideally if you had a small test scenario we could use to reproduce Note that we have some ruby integration tests for ruby [here](https://github.com/killbill/killbill-integration-tests/blob/master/killbill-integration-tests/core/usage/test_default_catalog_consumable_in_arrear.rb)


Otherwise, can you debug it and report on what you see for `startDate`, `endDate`, `targetDate` do you see? What do you have disk for already built usage? ...

**Comment by pierre** [comment]
@tytkal Did you have a chance to look at @sbrossie's questions?

**Comment by tytkal** [comment]
@pierre @sbrossie  Yes I have look to it, I will debug it after one or two weeks.

**Comment by sbrossie** [comment]
@tytkal Any update, or should we close the issue?

**Comment by tytkal** [comment]
we still have this issue. 
I have debugged it and the transitionTimes was empty, the value of startdate is 16-10-2016 and enddate is 16-10-2016 and targetDate is 26-01-2017
attached screen shot of the debugger values.

![image](https://cloud.githubusercontent.com/assets/264391/25705941/a5bd3994-30e7-11e7-8c61-88aa08f3ffc6.png)

**Comment by sbrossie** [comment]
@tytkal Ok i looked again at the code, and i can see how this is happening but i either need you to debug it and and help me verify my hypothesis or provide me with a test scenario i could run on my end.

In your case, the `transitionTimes` ends up being an empty list, and what i would need are the values for the following variables (or a test case to reproduce):
* `rawUsageStartDate`
* `closedInterval`
* `startDate`
* `endDate`
* `billingEvents`

Actually maybe an easier way would be to export your data using our [export tool](https://github.com/killbill/killbill/wiki/Production-troubleshooting#exporting-account-data) and send it over to me so i can look into it.

**Comment by tytkal** [comment]
@sbrossie I have manage to write test case that have same value. and it is crash 
`    public void testComputeMissingItems() throws CatalogApiException {

        final LocalDate startDate = new LocalDate(2014, 10, 16);
        final LocalDate firstBCDDate = new LocalDate(2014, 04, 15);
        final LocalDate endDate = new LocalDate(2014, 10, 16);

        // 2 items for startDate - firstBCDDate
        final List<RawUsage> rawUsages = new ArrayList<RawUsage>();
        rawUsages.add(new DefaultRawUsage(subscriptionId, new LocalDate(2014, 10, 26), "unit", 130L));
        rawUsages.add(new DefaultRawUsage(subscriptionId, new LocalDate(2014, 03, 21), "unit", 271L));
        // 1 items for firstBCDDate - endDate
        rawUsages.add(new DefaultRawUsage(subscriptionId, new LocalDate(2014, 04, 15), "unit", 199L));

        final DefaultTieredBlock block = createDefaultTieredBlock("unit", 100, 10, BigDecimal.ONE);
        final DefaultTier tier = createDefaultTier(block);
        final DefaultUsage usage = createDefaultUsage(usageName, BillingPeriod.MONTHLY, tier);

        final LocalDate targetDate = endDate;

        final BillingEvent event1 = createMockBillingEvent(new LocalDate(2014, 10, 16).toDateTimeAtStartOfDay(DateTimeZone.UTC),BillingPeriod.MONTHLY, Collections.<Usage>emptyList());
        final BillingEvent event2 = createMockBillingEvent(new LocalDate(2014, 10, 16).toDateTimeAtStartOfDay(DateTimeZone.UTC), BillingPeriod.MONTHLY, Collections.<Usage>emptyList());

        final AccountDateAndTimeZoneContext accountDateAndTimeZoneContext = new DefaultAccountDateAndTimeZoneContext(clock.getUTCNow(), internalCallContext);
        final ContiguousIntervalConsumableInArrear intervalConsumableInArrear = new ContiguousIntervalConsumableInArrear(usage, accountId, invoiceId, rawUsages, targetDate, new LocalDate(2014, 10, 26), accountDateAndTimeZoneContext);
            intervalConsumableInArrear.addBillingEvent(event1);
            intervalConsumableInArrear.addBillingEvent(event2);
        intervalConsumableInArrear.build(true);
`
I can push it in other branch if you want.

**Comment by sbrossie** [comment]
@tytkal I see how that test make the problem appear, but what i am unclear, is how does that map to a real case scenario?

* Your `startDate`, `endDate` and `targetDate` are all set to `2014-10-16`. I suppose that would corresponds to a use case where subscription was created and then cancelled on the same day.So far so good...
* Your `rawUsageStartDate` (optimization date generated by code to not look at things too far in the past is actually ahead of `startDate`, `endDate` and `targetDate` and as a result the `transitionTimes` list is empty, creating the NPE.

So while we could fix the code (easily) to avoid generating the NPE, i am still not clear how that maps to a real use case?

**Comment by tytkal** [comment]
@sbrossie OK I think I have a clue how this happen, Our event system was down a while and all the massages (usage report ) was stuck and not delivered, so may be that why we have  usage with this date

**Comment by sbrossie** [comment]
@tytkal I reused your test case and submitted a fix to avoid the NPE, so i think that should address the issue. However the fix is only on 0.18.x release (0.16.x is not updated unless there is a blinking red issue).
